internal/ethapi/api_test.go

```diff
@@ -0,0 +1,645 @@
+			},
+			overrides: StateOverride{
+				randomAccounts[2].addr: OverrideAccount{
+					Code:      hex2Bytes("6080604052348015600f57600080fd5b506004361060285760003560e01c80638381f58a14602d575b600080fd5b60336049565b6040518082815260200191505060405180910390f35b6000548156fea2646970667358221220eab35ffa6ab2adfe380772a48b8ba78e82a1b820a18fcb6f59aa4efb20a5f60064736f6c63430007040033"),
+					StateDiff: &map[common.Hash]common.Hash{common.Hash{}: common.BigToHash(big.NewInt(123))},
+				},
+			},
+			want: "0x000000000000000000000000000000000000000000000000000000000000007b",
+		},
+		// TODO add testcase exhausting gas cap
+	}
+	for i, tc := range testSuite {
+		result, err := api.Call(context.Background(), tc.call, rpc.BlockNumberOrHash{BlockNumber: &tc.blockNumber}, &tc.overrides)
+		if tc.expectErr != nil {
+			if err == nil {
+				t.Errorf("test %d: want error %v, have nothing", i, tc.expectErr)
+				continue
+			}
+			if !errors.Is(err, tc.expectErr) {
+				// Second try
+				if !reflect.DeepEqual(err, tc.expectErr) {
+					t.Errorf("test %d: error mismatch, want %v, have %v", i, tc.expectErr, err)
+				}
+			}
+			continue
+		}
+		if err != nil {
+			t.Errorf("test %d: want no error, have %v", i, err)
+			continue
+		}
+		if !reflect.DeepEqual(result.String(), tc.want) {
+			t.Errorf("test %d, result mismatch, have\n%v\n, want\n%v\n", i, result.String(), tc.want)
+		}
+	}
+}
+
+func TestBatchCall(t *testing.T) {
+	t.Parallel()
+	// Initialize test accounts
+	var (
+		accounts  = newAccounts(3)
+		genBlocks = 10
+		signer    = types.HomesteadSigner{}
+		genesis   = &core.Genesis{
+			Config: params.TestChainConfig,
+			Alloc: core.GenesisAlloc{
+				accounts[0].addr: {Balance: big.NewInt(params.Ether)},
+				accounts[1].addr: {Balance: big.NewInt(params.Ether)},
+				accounts[2].addr: {Balance: big.NewInt(params.Ether)},
+			},
+		}
+	)
+	api := NewBlockChainAPI(newTestBackend(t, genBlocks, genesis, func(i int, b *core.BlockGen) {
+		// Transfer from account[0] to account[1]
+		//    value: 1000 wei
+		//    fee:   0 wei
+		tx, _ := types.SignTx(types.NewTx(&types.LegacyTx{
+			Nonce:    uint64(i),
+			To:       &accounts[1].addr,
+			Value:    big.NewInt(1000),
+			Gas:      params.TxGas,
+			GasPrice: b.BaseFee(),
+			Data:     nil,
+		}), signer, accounts[0].key)
+		b.AddTx(tx)
+	}))
+	var (
+		randomAccounts = newAccounts(3)
+		latestBlock    = rpc.LatestBlockNumber
+	)
+	type res struct {
+		Return string
+		Error  string
+	}
+	var testSuite = []struct {
+		config    BatchCallConfig
+		expectErr error
+		want      []res
+	}{
+		// First value transfer OK after state override, second one should succeed
+		// because of first transfer.
+		{
+			config: BatchCallConfig{
+				Block: rpc.BlockNumberOrHash{BlockNumber: &latestBlock},
+				StateOverrides: &StateOverride{
+					randomAccounts[0].addr: OverrideAccount{Balance: newRPCBalance(big.NewInt(1000))},
+				},
+				Calls: []BatchCallArgs{{
+					TransactionArgs: TransactionArgs{
```

Comment:
`randomAccounts[0].addr` and `randomAccounts[1].addr` are all have funds to transfer(allocated in genesis), you should use a new address here.